grin/src/Transformations/ExtendedSyntax/Optimising/NonSharedElimination.hs

```diff
@@ -0,0 +1,36 @@
+{-# LANGUAGE LambdaCase, RecordWildCards #-}
+module Transformations.ExtendedSyntax.Optimising.NonSharedElimination where
+
+{-
+Remove the updates that update only non-shared locations.
+-}
+
+import qualified Data.Set as Set
+import Data.Functor.Foldable as Foldable
+
+import Grin.ExtendedSyntax.Grin
+import Grin.ExtendedSyntax.TypeEnv (ptrLocations)
+import Grin.ExtendedSyntax.TypeCheck (typeEnvFromHPTResult)
+import Transformations.ExtendedSyntax.Names (ExpChanges(..))
+import AbstractInterpretation.ExtendedSyntax.Sharing.Result (SharingResult(..))
+
+
+
+nonSharedElimination :: SharingResult -> Exp -> (Exp, ExpChanges)
+nonSharedElimination SharingResult{..} exp = (exp', change) where
+
+  exp' = cata skipUpdate exp
+
+  change = if exp' /= exp then DeletedHeapOperation else NoChange
+
+  tyEnv = either error id $ typeEnvFromHPTResult _hptResult
```

Comment:
This is a bit fishy. We should not convert HPT results in the transformation. :) Something to worth looking at when we clean the pipeline code.

Reply:
Is this an impossible case?

Reply:
@andorp Yep, there is something wrong about this. I can't recall the reason why I implemented the change. I will create an issue about this.
@csabahruska Well, if the sharing result is correct, we should always get a valid HPT results ad thus a valid type env. This is basically a safety catch.

Reply:
See #95